pkg/volume/csi/csi_plugin.go

```diff
@@ -305,15 +343,17 @@ func (p *csiPlugin) NewMounter(
 
 	// persist volume info data for teardown
 	node := string(p.host.GetNodeName())
-	attachID := getAttachmentName(pvSource.VolumeHandle, pvSource.Driver, node)
 	volData := map[string]string{
-		volDataKey.specVolID:    spec.Name(),
-		volDataKey.volHandle:    pvSource.VolumeHandle,
-		volDataKey.driverName:   pvSource.Driver,
-		volDataKey.nodeName:     node,
-		volDataKey.attachmentID: attachID,
+		volDataKey.specVolID:  spec.Name(),
+		volDataKey.volHandle:  volumeHandle,
+		volDataKey.driverName: driverName,
+		volDataKey.nodeName:   node,
+		volDataKey.driverMode: string(driverMode),
```

Comment:
are we going to have issues adding/dropping fields from the `volData` object between Kubernetes versions? Just modifying the jsons like this will cause data incompatibility (maybe crashes). The answer may be that we don't care since we're probably draining the nodes on upgrades but just want to make sure.

If we want to support files produced by old Kubelet in the future we may want to consider using protobuf and use proto best practices to ensure backwards/forwards support

@msau42 @saad-ali @jsafrane

Reply:
It is only an addition of field here (`volData[driverMode]`). Nothing is removed to avoid exactly the issue you stated. Older k8s will not ask for a missing field.